### PR TITLE
Add puppet-naginator replacement

### PR DESCRIPTION
**Pull request by nightlyone**
in order to do in 3s what puppet-naginator does in > 20m (speedup ~400x)

### CODE REVIEW

**Comment by nightlyone** [comment]
any feedback here?

**Comment by fungusakafungus** [comment]
What does it exactly do and how is it supposed to replace naginator? Like, the general workflow.

I see two new binaries, `puppet-naginator` and `collect-resources`. When are those supposed to be called?

**Comment by mlafeldt** [comment]
> What does it exactly do and how is it supposed to replace naginator? Like, the general workflow.

Like to know that as well.

**Comment by nightlyone** [comment]
`puppet-naginator` queries puppetdb for collected resources of all nagios types and writes out the result to stdout.  So the definition of resources to collect would not change.

But it could for example replace 
```puppet
Nagios_service <<||>> {
  target  => '/etc/nagios3/conf.d/collected_services.cfg',
  notify  => Exec['fix-permissions'],
  require => [Package['nagios3'],File['/etc/nagios3/conf.d']],
}
```
in the configuration for the nagios server at https://github.com/Jimdo/configuration-management/blob/production/dist/nagios/manifests/server.pp with
```puppet
exec { 'nagios_collect_services'}
  command => "puppet-naginator -t service > /etc/nagios3/conf.d/collected_services.cfg",
  notify  => Exec['fix-permissions'],
  require => [Package['nagios3'],File['/etc/nagios3/conf.d']],
}
```

in order to save a lot of cpu time and allow e.g. more servers or shorter puppet run intervals.

The `collect-resources` binary allows to add collected resources to other services as well. It is technically not necessary, but useful for debugging and as stepping stone to solve this bottleneck elsewhere, too.

I would remove `collect-resources` from the changeset, if it is confusing.

**Comment by fungusakafungus** [comment]
cool. Sounds great! So what are TODOs here?

File: cmd/puppet-naginator/templates.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	"strings"
+	"time"
+
+	"github.com/nightlyone/puppetquery"
+)
+
+const nagiosPrefix = "Nagios_"
+
+var keyOrder = slice2map([]string{"host_name", "alias", "use", "parents", "service_description"})
+var badparams = slice2map([]string{"notify", "target", "ensure", "require", "before", "tag"})
```
Comment (fungusakafungus) [inline cmd/puppet-naginator/templates.go]: those are called metaparameter in puppet. Except target. I don't know target. Here's a list: https://docs.puppetlabs.com/references/latest/metaparameter.html

Comment (nightlyone) [inline cmd/puppet-naginator/templates.go]: incorporated this in v0.6

**Pull request merged**